### PR TITLE
Creating Structures for Efficient Binary Classification Metrics and Binary Dice (4/5)

### PR DESCRIPTION
**Pull request by emersodb**
# PR Type

Feature

# Short Description

This PR is quite a bit larger. It creates the infrastructure needed for efficient binary classification metrics and implements an efficient binary dice metric using these structures. Some of the structures introduced might seem a bit unnecessary. For example the `ClassificationMetric` might seem like an unnecessary abstraction, but it will be used for the non-binary versions to come in a follow up PR.

# Tests Added

There are a lot of tests covering as many edge cases as I could think of both for the binary Dice metric and also for the efficient binary classification metric base that underlies the Dice.

### CODE REVIEW

File: fl4health/metrics/efficient_metrics.py [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+        denominator = denominator[denominator != 0]
+    else:
+        numerator[denominator == 0] = zero_division
+        denominator[denominator == 0] = 1
+
+    # Return individual dice coefficients
+    return numerator / denominator
+
+
+class BinaryDice(BinaryClassificationMetric):
```
Comment (scarere) [inline fl4health/metrics/efficient_metrics.py:49]: I'm trying to imagine in my head how multiclass dice would work. Because in theory if we have a label dimension for our preds and targets, then we could use BinaryDice? Would we get different values? What is the intuitive way of computing dice? how would people expect it to function?

File: fl4health/metrics/efficient_metrics_base.py [hunk 2]
```diff
@@ -0,0 +1,577 @@
...
+        self,
+        true_positives: torch.Tensor,
+        false_positives: torch.Tensor,
+        false_negatives: torch.Tensor,
+        true_negatives: torch.Tensor,
+    ) -> Metrics:
+        pass
+
+
+class BinaryClassificationMetric(ClassificationMetric):
```
[the remark this answers is not in the document]
Comment (scarere) [reply fl4health/metrics/efficient_metrics_base.py:321]: I'm wondering what this binary dice class is really adding besides the pos_label argument? Everything else seems like functionality not specific to binary metrics but needed for all classification metrics in general

**Review by scarere — COMMENTED** [review]
I think you did a great job of cleaning up the code here. There are a lot of changes that I think improve the overall clarity of the code but there are a few things that I find myself "missing" from the old implementation. Namely
- I'm partial to having a generic argument that allows the metrics to be computed along/over any of the dimensions as opposed to restricting it to just 2 dimensions (batch and label)
    - One improvement i like about having only 2 is that we can very explicitly state the shape and dimensionality of the incoming count tensors
- The BinaryClassificationMetric doesn't really seem to add a lot of functionality specific to binary metrics. Instead it seems to kind of just complete the functionality of the classification metric
    - I'm wondering if we can simplify whats going on in these base classes
- We no longer have the ability to ignore the background class in the base class.

Lastly one of the main problems we were having before was that the result returned by the metric was sometimes dependent whether or not your preds and targets had a label dimensions. It seems in this PR you resolved that by enforcing all preds and targets to be in range [0, 1]. Are we then assuming a label dimension always exists? Will we get the same result for binary classification task? Have we covered the edge cases here

**Comment by scarere** [comment]
> I think you did a great job of cleaning up the code here. There are a lot of changes that I think improve the overall clarity of the code but there are a few things that I find myself "missing" from the old implementation. Namely
> 
> * I'm partial to having a generic argument that allows the metrics to be computed along/over any of the dimensions as opposed to restricting it to just 2 dimensions (batch and label)
>   
>   * One improvement i like about having only 2 is that we can very explicitly state the shape and dimensionality of the incoming count tensors
> * The BinaryClassificationMetric doesn't really seem to add a lot of functionality specific to binary metrics. Instead it seems to kind of just complete the functionality of the classification metric
>   
>   * I'm wondering if we can simplify whats going on in these base classes
> * We no longer have the ability to ignore the background class in the base class.
> 
> Lastly one of the main problems we were having before was that the result returned by the metric was sometimes dependent whether or not your preds and targets had a label dimensions. It seems in this PR you resolved that by enforcing all preds and targets to be in range [0, 1]. Are we then assuming a label dimension always exists? Will we get the same result for binary classification task? Have we covered the edge cases here

**Comment by scarere** [comment]
Did we resolve the issue where metrics return different results based on whether the preds and targets have a label dimension. I can see that we're now requiring preds and targets to be in range [0, 1], but for binary tasks are we going to get the same result in both cases? Are there any checks in place or intuitive parts that can prevent the user from providing the inputs in the wrong form?

File: fl4health/metrics/efficient_metrics.py [hunk 3]
```diff
@@ -0,0 +1,145 @@
+from logging import WARNING
+
+import torch
+from flwr.common.logger import log
+from flwr.common.typing import Metrics, Scalar
+
+from fl4health.metrics.efficient_metrics_base import BinaryClassificationMetric
+
+
+def compute_dice_on_count_tensors(
```
[the remark this answers is not in the document]
Comment (scarere) [reply fl4health/metrics/efficient_metrics.py:10]: Yeah I guess it just seems strange to me that the method that actually computes the DICE score is not a part of the class. It's been awhile since I thought about the implementation for this PR so excuse me if i'm forgetting something, but intuitively it makes sense to me for BinaryMetrics to be a subclass of their Multiclass counterparts

[the remark this answers is not in the document]
Comment (emersodb) [reply fl4health/metrics/efficient_metrics.py:10]: Yeah I know what you mean. No worries. I'm only just able to get back to this after several weeks of my priorities being diverted. 

So the reason they are different is and we split them fairly heavily (similar to how tensorflow/keras does) is because of the differences that these types of counting based metrics have when considering a binary versus multi-class perspective. For the binary implementations, we are allowing for either vector encodings of predictions and targets (similar to the multiclass case) or non vector encoded (i.e. 0.8 implies [0.2, 0.8]). Because we're supporting that diversity of input, there are more sanity checks, post-processing steps etc. that need to be done for this setting that aren't necessary based on the restrictions we put on the multiclass case. They will share functionality in the base ClassificationMetric class, but they aren't properly subsets of one another the way we've chosen to go.

File: fl4health/metrics/efficient_metrics.py (hunk 1 above)
Comment (emersodb) [reply fl4health/metrics/efficient_metrics.py:49]: I've added more documentation to the class to make this more clear right at the start in the `__init__` docs specifically.

File: fl4health/metrics/efficient_metrics_base.py [hunk 4]
```diff
@@ -0,0 +1,577 @@
...
+from logging import WARNING
+
+import torch
+from flwr.common.logger import log
+from flwr.common.typing import Metrics, Scalar
+
+from fl4health.metrics.metrics_base import Metric
+
+
+class MetricOutcome(Enum):
```
[the remark this answers is not in the document]
Comment (emersodb) [reply fl4health/metrics/efficient_metrics_base.py:12]: I guess I think of this as an Outcome from a prediction. That is, we predicted a class label and it was correct (TP), we didn't predict a class label and were wrong (FP). That's why I used the work outcome. 

I think using "confusion" is a bit too specific, since these have value outside of confusion matrices?

Perhaps `OutcomeCounts`? I thought maybe `PredictionCounts` or `ClassificationCounts` but that sort of implies the counts are just related to prediction stats.

File: fl4health/metrics/efficient_metrics_base.py [hunk 5]
```diff
@@ -0,0 +1,577 @@
...
+    FALSE_POSITIVE = "false_positive"
+    TRUE_NEGATIVE = "true_negative"
+    FALSE_NEGATIVE = "false_negative"
+
+
+class ClassificationMetric(Metric, ABC):
+    def __init__(
+        self,
+        name: str,
+        label_dim: int | None,
```
[the remark this answers is not in the document]
Comment (emersodb) [reply fl4health/metrics/efficient_metrics_base.py:23]: Because this forms an abstract base class, it is acceptable that the variable not be used in the class itself. This serves to enforce an interface for the `__init__` function and also to provide each inheriting class access to this variable.

For the second point. I think that's fair. We are specifying the inheriting classes and have full control over the expected functionality. So I was erring on the side of having more documentation than less. Technically any method could be overridden in an inheriting class rendering any docs in this class moot. I'll do a little bit of sparsification of the docs, but I don't want to remove too much if I think it's helpful in understanding the point of the class.

File: fl4health/metrics/efficient_metrics_base.py [hunk 6]
```diff
@@ -0,0 +1,577 @@
...
+        batch_dim: int | None,
```
[the remark this answers is not in the document]
Comment (emersodb) [reply fl4health/metrics/efficient_metrics_base.py:24]: You and I talked about this a bit together at one point and, if I recall correctly, we were able to come up with a scenario or two where someone might want the flexibility that you're describing. However, I think the **predominant** use cases are covered by allowing for a batch and label dimension to be specified if it exists (binary problems might not have one). 

I think these implementations are much clearer when we provide names to the dimensions and are able to concretely talk about what they mean in context and what the resulting implications are in specifying them. Since we're not aiming to build a generic and fully functional metrics library within the FL library, my take is that this is already general enough to hit 80-90% of use cases and if someone wants something different they can implement it using the scaffolding or idea.

File: fl4health/metrics/efficient_metrics_base.py [hunk 7]
```diff
@@ -0,0 +1,577 @@
...
+        If the predictions provided are continuous in value, then the associated counts will also be continuous
+        ("soft"). For example, with a target of 1, a prediction of 0.8 contributes 0.8 to the true_positives count and
+        0.2 to the false_negatives.
+
+        NOTE: Preds and targets are expected to have elements in the interval [0, 1] or to be thresholded, using
+        the argument of this class to be as such.
+
+        Args:
+            name (str): The name of the metric.
+            label_dim (int | None, optional): Specifies which dimension in the provided tensors corresponds to the
```
[the remark this answers is not in the document]
Comment (emersodb) [reply fl4health/metrics/efficient_metrics_base.py:52]: Agreed. The term that I have settled on (and am open to changing) is that targets or predictions are "vector encoded" when they have an element for each class in a problem. This is true, even for binary problems, where you might have a two-element vector [0.8, 0.2] corresponding to the negative and positive classes. I have used "class encoded" or not "vector encoded" to mean that the class index is being predicted for example [3, 2, 6, 1] for a batch size of 4 and class index predictions of 3, 2, 6, and 1 respectively. I may also use index encoded, but I'll try to look for that in the code and use class encoded uniformly.

File: fl4health/metrics/efficient_metrics_base.py [hunk 8]
```diff
@@ -0,0 +1,577 @@
...
+        self.counts_initialized = False
+        self.true_positives, self.false_positives, self.false_negatives, self.true_negatives = (
+            torch.tensor([]),
+            torch.tensor([]),
+            torch.tensor([]),
+            torch.tensor([]),
+        )
+
+    @classmethod
+    def _threshold_tensor(cls, input: torch.Tensor, threshold: float | int) -> torch.Tensor:
```
[the remark this answers is not in the document]
Comment (emersodb) [reply fl4health/metrics/efficient_metrics_base.py:120]: Yeah I think that makes sense. I'm already creating a utility file so I'll drop this in there instead.

File: fl4health/metrics/efficient_metrics_base.py [hunk 9]
```diff
@@ -0,0 +1,577 @@
...
+
+        # On the off chance were given booleans convert them to integers
+        preds = preds.to(torch.uint8) if preds.dtype == torch.bool else preds
+        targets = targets.to(torch.uint8) if targets.dtype == torch.bool else targets
+
+        # Maybe threshold predictions into 'hard' predictions.
+        preds = preds if self.threshold is None else self._threshold_tensor(preds, self.threshold)
+        return preds, targets
+
+    def _assert_correct_ranges(self, preds: torch.Tensor, targets: torch.Tensor) -> None:
```
[the remark this answers is not in the document]
Comment (emersodb) [reply fl4health/metrics/efficient_metrics_base.py:175]: This is used for the `BinaryClassificationMetric` here and also in the `MultiClassificationMetric` in a future PR. So rather than repeating the functionality I centralized it here. To answer your question, for the `BinaryClassificationMetric`, we are supporting preds/targets without a label dimension because it might be implied i.e. a 1D prediction tensor of length batch size. For  `MultiClassificationMetric`, we are going to require a label dimension to make our lives easier.

This function is just mean to be run after things like thresholding have been completed to make sure everything is in a form that makes sense for metric computation.

File: fl4health/metrics/efficient_metrics_base.py [hunk 10]
```diff
@@ -0,0 +1,577 @@
...
+        Ensures that the prediction and target tensors are within the expected ranges for computation.
+
+        Args:
+            preds (torch.Tensor): Predictions tensor
+            targets (torch.Tensor): Targets tensor
+        """
+        assert torch.min(preds) >= 0 and torch.max(preds) <= 1, "Expected preds to be in range [0, 1]."
+        assert torch.min(targets) >= 0 and torch.max(targets) <= 1, "Expected targets to be in range [0, 1]."
+
+    def _prepare_counts_from_preds_and_targets(
```
[the remark this answers is not in the document]
Comment (emersodb) [reply fl4health/metrics/efficient_metrics_base.py:186]: This is actually used in `count_tp_fp_fn_tn` for both this `BinaryClassificationMetric` and `MultiClassificationMetric`classes, hence its centralization in this base class here. This is the downside of splitting PRs into smaller chunks. Sometimes the reason for an implementation like this isn't clear until you get to the next PR 😢

File: fl4health/metrics/efficient_metrics_base.py [hunk 11]
```diff
@@ -0,0 +1,577 @@
...
+                false negative, and true negative indications for each of the predictions in the provided tensors.
+        """
+
+        true_positives = (preds * targets) if not self.discard_tp else torch.tensor([])
+        false_positives = (preds * (1 - targets)) if not self.discard_fp else torch.tensor([])
+        false_negatives = ((1 - preds) * targets) if not self.discard_fn else torch.tensor([])
+        true_negatives = ((1 - preds) * (1 - targets)) if not self.discard_tn else torch.tensor([])
+        return true_positives, false_positives, false_negatives, true_negatives
+
+    def _sum_along_axes_or_discard(
```
[the remark this answers is not in the document]
Comment (emersodb) [reply fl4health/metrics/efficient_metrics_base.py:207]: As with the implementation of `_prepare_counts_from_preds_and_targets` this functionality is called in the `count_tp_fp_fn_tn` methods of both the `BinaryClassificationMetric` and `MultiClassificationMetricclasses`. So I centralized it here. 🙂

File: fl4health/metrics/efficient_metrics_base.py [hunk 12]
```diff
@@ -0,0 +1,577 @@
...
+            self.counts_initialized = True
+            return
+
+        # If batch_dim has been specified, we accumulate counts for EACH INSTANCE seen throughout the updates, such
+        # that each of the counts is a 2D tensor of row length equal to the samples seen.
+        # NOTE: This ASSUMES the batch dimension comes FIRST for the counts.
+        # Otherwise, the counts are 1D tensors with length equal to the number of classes (or possibly 1 if using the
+        # BinaryClassificationMetric)
+        self.true_positives = (
+            torch.cat([self.true_positives, tp], dim=0) if self.batch_dim is not None else self.true_positives + tp
```
[the remark this answers is not in the document]
Comment (emersodb) [reply fl4health/metrics/efficient_metrics_base.py:260]: That is a fair point. This function **assumes** that the inheriting class implementation of `count_tp_fp_fn_tn` will always return counts where the batch dimension is the first dimension when `batch_dim` is defined. You can see the note on line 256. I will add a note to the functions documentation as well to make this a little more clear. This base class is tightly coupled to the implementing classes of `BinaryClassificationMetric` and `MultiClassificationMetricclasses`. The goal is that it is quite rare that someone would be inheriting from this class but rather inheriting from those two child classes to create their own metrics. So I'm making some strong assumptions about the computations because I control them in the children.

File: fl4health/metrics/efficient_metrics_base.py [hunk 13]
```diff
@@ -0,0 +1,577 @@
...
+        """
+        Reset accumulated tp, fp, fn and tn's. They will be initialized with correct shape on next update
+        """
+        self.true_positives = torch.tensor([])
+        self.false_positives = torch.tensor([])
+        self.false_negatives = torch.tensor([])
+        self.true_negatives = torch.tensor([])
+        self.counts_initialized = False
+
+    def compute(self, name: str | None = None) -> Metrics:
```
[the remark this answers is not in the document]
Comment (emersodb) [reply fl4health/metrics/efficient_metrics_base.py:282]: The reason we're defining the return value as Metrics here is because that aligns with the return value expected by Flower for things like the client EvaluateRes and FitRes aliases. Restricting to a smaller set of them makes some sense, but also makes our lives a bit harder because of the typing misalignment etc. 

As a side note, the Scalar type now allows `List[T]` for each of these primitive types as well in the updated version of Flower that we've moved to...

File: fl4health/metrics/efficient_metrics_base.py [hunk 14]
```diff
@@ -0,0 +1,577 @@
...
+            false_positives=self.false_positives,
+            false_negatives=self.false_negatives,
+            true_negatives=self.true_negatives,
+        )
+        if name is not None:
+            metrics = {f"{name} - {k}": v for k, v in metrics.items()}
+        return metrics
+
+    @abstractmethod
+    def count_tp_fp_fn_tn(
```
[the remark this answers is not in the document]
Comment (emersodb) [reply fl4health/metrics/efficient_metrics_base.py:305]: I do, in fact, do this for the `BinaryClassificationMetric` (and the `MultiClassificationMetric` in a follow up PR) which are meant to be the classes from which a user inherits to define their metrics (as I do with the `BinaryDice`)

File: fl4health/metrics/efficient_metrics_base.py [hunk 15]
```diff
@@ -0,0 +1,577 @@
...
+
+    @abstractmethod
+    def compute_from_counts(
+        self,
+        true_positives: torch.Tensor,
+        false_positives: torch.Tensor,
+        false_negatives: torch.Tensor,
+        true_negatives: torch.Tensor,
+    ) -> Metrics:
+        pass
```
[the remark this answers is not in the document]
Comment (emersodb) [reply fl4health/metrics/efficient_metrics_base.py:318]: Since technically the `__call__` method isn't necessary, I didn't want to force this for anything inheriting from this class. However, I do, in fact, force an implementation if one wants to use it for convenience by leaving it as throwing a `NotImplementedError` in the `BinaryClassificationMetric` child class. So if you call it and haven't defined it, you get a slap on the write 😂

File: fl4health/metrics/efficient_metrics_base.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (emersodb) [reply fl4health/metrics/efficient_metrics_base.py:321]: In the current implementation, there is no dissonance as to whether the metrics are computed over two classes or a "single" class. Counts are **always** reported with respect to the class represented by the pos_label index. That is, you will never get situations where you're averaging between metrics isolated to both the negative and positive classes. The only thing that might happen is you are averaging over samples (i.e accuracy for pixels in each image averaged) or combining all counts (i.e. accuracy for all pixels in the dataset as a hold).

Note that this isn't a binary dice class. This is a more generic binary classification class where we're counting outcomes for a binary problem. This class is fundamentally different from that of the multiclass case (which is introduced in a subsequent PR) in that it allows for you to have either vector encoded preds/targets or implied preds/targets (i.e. 0.8 corresponds to a [0.2, 0.8] prediction in reality). The multi-class case will not allow this. Moreover, this class **forces** the reporting to be specific to the pos_label index as outlined above. If you don't want that behavior, you have to use the multiclass setting.

File: fl4health/metrics/efficient_metrics_base.py [hunk 16]
```diff
@@ -0,0 +1,577 @@
...
+        """
+        super().__init__(
+            name=name,
+            dtype=dtype,
+            label_dim=label_dim,
+            batch_dim=batch_dim,
+            threshold=threshold,
+            discard=discard,
+        )
+        assert pos_label == 0 or pos_label == 1, "pos_label must be either 0 or 1"
```
[the remark this answers is not in the document]
Comment (emersodb) [reply fl4health/metrics/efficient_metrics_base.py:403]: This implementation follows that of the implementations like sklearn. What it's allowing is for the user to say they want to see the metric with respect to the specified index. For example, this would allow you to see the recall associated with the positive label or the negative label separately. They aren't necessarily equivalents of one another. Similarly you could see the negative or positive class F1 score if desired.

File: fl4health/metrics/efficient_metrics_base.py [hunk 17]
```diff
@@ -0,0 +1,577 @@
...
+        self.pos_label = pos_label
+
+    def _transform_tensors(self, preds: torch.Tensor, targets: torch.Tensor) -> tuple[torch.Tensor, torch.Tensor]:
```
[the remark this answers is not in the document]
Comment (emersodb) [reply fl4health/metrics/efficient_metrics_base.py:406]: This is a good question and again highlights a challenge associated with breaking up PRs. This function is different in the multiclass setting. So having it be something that is overridden makes sense from a code sharing perspective if you look at the `MultiClassificationMetric` implementation, but makes a little less sense here 😢

File: fl4health/metrics/efficient_metrics_base.py [hunk 18]
```diff
@@ -0,0 +1,577 @@
...
+
+        return preds, targets
+
+    def _post_process_count_tensor(self, count_tensor: torch.Tensor) -> torch.Tensor:
+        """
+        Given a count tensor, in the various forms that it might appear, we need to post process these so that they
+        can be returned to the user in the appropriate format. The structure of the counts tensors after processing
+        a batch of data will differ depending on whether the label_dim and batch_dim have been specified.
+
+        NOTE: If the count has been specified as discarded, it will always simply be an empty tensor
```
[the remark this answers is not in the document]
Comment (emersodb) [reply fl4health/metrics/efficient_metrics_base.py:430]: That is one of the benefits, in addition to them having very concrete definitions to a user...i.e. tell me these dimensions and we're good to go. The more generic setting is somewhat nice but makes life much more difficult to cover a few more cases. Since we're not building a generic metrics library, I think it makes sense to have some simplification.

File: fl4health/metrics/efficient_metrics_base.py [hunk 19]
```diff
@@ -0,0 +1,577 @@
...
+
+        Returns:
+            tuple[torch.Tensor, torch.Tensor]: Potentially transformed predictions and targets tensors, in that order
+        """
+        preds, targets = super()._transform_tensors(preds, targets)
+        super()._assert_correct_ranges(preds, targets)
+
+        return preds, targets
+
+    def _post_process_count_tensor(self, count_tensor: torch.Tensor) -> torch.Tensor:
```
[the remark this answers is not in the document]
Comment (emersodb) [reply fl4health/metrics/efficient_metrics_base.py:424]: Unfortunately, In order for us to be consistent in terms of computation, this reshaping kind of has to happen somewhere. This reshaping is a direct result of the finite, but various, ways in which someone can define a binary prediction problem in terms of pred/target shapes. It's annoying, but if we want to support users in these representations, we need to "handle" them somehow. So I chose to do it this way, which is ugly but effective.

File: fl4health/metrics/efficient_metrics_base.py [hunk 20]
```diff
@@ -0,0 +1,577 @@
...
+                count_tensor = count_tensor.transpose(0, 1)
+
+            if count_tensor.shape[1] == 2:
+                # Always return the class with label 1 (pos_label 0 will be handled by rearranging counts if necessary)
+                return count_tensor[:, 1].unsqueeze(1)
+            elif count_tensor.shape[1] == 1:
+                return count_tensor
+            else:
+                raise ValueError(f"Label dimension has unexpected size of {count_tensor.shape[1]}")
+        elif self.batch_dim is not None:
```
[the remark this answers is not in the document]
Comment (emersodb) [reply fl4health/metrics/efficient_metrics_base.py:473]: This is handled within the else statement below on Line 479. There are two scenarios handled by the else. 

The first is the case that you mention, where `label_dim` is not None and `batch_dim` is None. In this case, the `count_tensor` will have either 2 or 1 elements (1D tensor). The two different scenarios are because you might have preds vector encoded (two elements per prediction, one for each label) or the preds are implied (i.e. 0.8 to represent [0.2, 0.8]), respectively.

The second is the case where both `label_dim` and `batch_dim` are both None. In this case, everything should have collapsed to a single value (because the labels are assumed to be implicit when label dims are not specified).

I think the way I did this makes that a bit unclear. So I will make it more explicit in the way the cases are handled.

**Comment by emersodb** [comment]
> Did we resolve the issue where metrics return different results based on whether the preds and targets have a label dimension. I can see that we're now requiring preds and targets to be in range [0, 1], but for binary tasks are we going to get the same result in both cases? Are there any checks in place or intuitive parts that can prevent the user from providing the inputs in the wrong form?

Short answer is yes 🙂. The binary class here will only ever give you a result with respect to one of your label classes (as specified by `pos_label`)

File: fl4health/metrics/efficient_metrics_base.py [hunk 21]
```diff
@@ -0,0 +1,577 @@
...
+        return metrics
+
+    @abstractmethod
+    def count_tp_fp_fn_tn(
+        self, preds: torch.Tensor, targets: torch.Tensor
+    ) -> tuple[torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor]:
+        pass
+
+    @abstractmethod
+    def compute_from_counts(
```
[the remark this answers is not in the document]
Comment (emersodb) [reply fl4health/metrics/efficient_metrics_base.py:270]: Done!

**Review by scarere — APPROVED** [review]
